From ba5a944f3eef890b89c8e79b3c0c01edc4a8ffd2 Mon Sep 17 00:00:00 2001
From: Amith A <quic_amitajit@quicinc.com>
Date: Mon, 25 Mar 2024 19:45:33 +0530
Subject: [PATCH] wifi: nl80211: Add NL attribute to denote ft_auth sta flag

A few drivers may need to know if STA is using FT auth to decide on
allowing set key.

Add new netlink attribute NL80211_STA_FLAG_FT_AUTH to get the ft_auth
sta flag. This flag is used to set need_ptk_4_way to false so that the
driver is notified about the FT auth and allows set key.

Signed-off-by: Amith A <quic_amitajit@quicinc.com>
---
 include/uapi/linux/nl80211.h |  1 +
 net/wireless/nl80211.c       | 10 ++++++----
 2 files changed, 7 insertions(+), 4 deletions(-)

--- a/include/uapi/linux/nl80211.h
+++ b/include/uapi/linux/nl80211.h
@@ -3607,6 +3607,7 @@ enum nl80211_iftype {
  * @NL80211_STA_FLAG_ASSOCIATED: station is associated; used with drivers
  *	that support %NL80211_FEATURE_FULL_AP_CLIENT_STATE to transition a
  *	previously added station into associated state
+ * @NL80211_STA_FLAG_FT_AUTH: station uses FT authentication
  * @NL80211_STA_FLAG_MAX: highest station flag number currently defined
  * @__NL80211_STA_FLAG_AFTER_LAST: internal use
  */
@@ -3619,6 +3620,7 @@ enum nl80211_sta_flags {
 	NL80211_STA_FLAG_AUTHENTICATED,
 	NL80211_STA_FLAG_TDLS_PEER,
 	NL80211_STA_FLAG_ASSOCIATED,
+	NL80211_STA_FLAG_FT_AUTH,
 
 	/* keep last */
 	__NL80211_STA_FLAG_AFTER_LAST,
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -7627,7 +7627,7 @@ int cfg80211_check_station_change(struct
 		return -EINVAL;
 
 	/* When you run into this, adjust the code below for the new flag */
-	BUILD_BUG_ON(NL80211_STA_FLAG_MAX != 7);
+	BUILD_BUG_ON(NL80211_STA_FLAG_MAX != 8);
 
 	switch (statype) {
 	case CFG80211_STA_MESH_PEER_KERNEL:
@@ -7710,7 +7710,8 @@ int cfg80211_check_station_change(struct
 				  BIT(NL80211_STA_FLAG_ASSOCIATED) |
 				  BIT(NL80211_STA_FLAG_SHORT_PREAMBLE) |
 				  BIT(NL80211_STA_FLAG_WME) |
-				  BIT(NL80211_STA_FLAG_MFP)))
+				  BIT(NL80211_STA_FLAG_MFP) |
+				  BIT(NL80211_STA_FLAG_FT_AUTH)))
 			return -EINVAL;
 
 		/* but authenticated/associated only if driver handles it */
@@ -8302,7 +8303,7 @@ static int nl80211_new_station(struct sk
 		return -EINVAL;
 
 	/* When you run into this, adjust the code below for the new flag */
-	BUILD_BUG_ON(NL80211_STA_FLAG_MAX != 7);
+	BUILD_BUG_ON(NL80211_STA_FLAG_MAX != 8);
 
 	switch (dev->ieee80211_ptr->iftype) {
 	case NL80211_IFTYPE_AP:
@@ -8318,7 +8319,8 @@ static int nl80211_new_station(struct sk
 		    info->attrs[NL80211_ATTR_PEER_AID])
 			return -EINVAL;
 		/* but don't bother the driver with it */
-		params.sta_flags_mask &= ~BIT(NL80211_STA_FLAG_TDLS_PEER);
+		params.sta_flags_mask &= ~(BIT(NL80211_STA_FLAG_TDLS_PEER) |
+					   BIT(NL80211_STA_FLAG_FT_AUTH));
 
 		/* allow authenticated/associated only if driver handles it */
 		if (!(rdev->wiphy.features &
